crates/bevy_reflect/src/reflect.rs

```diff
@@ -215,19 +212,67 @@ pub trait Reflect: Any + Send + Sync {
     }
 }
 
-impl Debug for dyn Reflect {
+impl Debug for dyn PartialReflect {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         self.debug(f)
     }
 }
 
-impl Typed for dyn Reflect {
+impl Typed for dyn PartialReflect {
     fn type_info() -> &'static TypeInfo {
         static CELL: NonGenericTypeInfoCell = NonGenericTypeInfoCell::new();
         CELL.get_or_set(|| TypeInfo::Value(ValueInfo::new::<Self>()))
     }
 }
 
+impl dyn PartialReflect {
+    /// Downcasts the value to type `T`, consuming the trait object.
+    ///
+    /// If the underlying value does not implement [`Reflect`]
+    /// or is not of type `T`, returns `Err(self)`.
+    pub fn try_downcast<T: Reflect>(
+        self: Box<dyn PartialReflect>,
+    ) -> Result<Box<T>, Box<dyn PartialReflect>> {
+        self.into_full()?
+            .downcast()
+            .map_err(PartialReflect::into_partial)
+    }
+    /// Downcasts the value to type `T`, unboxing and consuming the trait object.
+    ///
+    /// If the underlying value does not implement [`Reflect`]
+    /// or is not of type `T`, returns `Err(self)`.
+    pub fn try_take<T: Reflect>(
+        self: Box<dyn PartialReflect>,
+    ) -> Result<T, Box<dyn PartialReflect>> {
+        self.try_downcast().map(|value| *value)
+    }
+
+    /// Downcasts the value to type `T` by reference.
+    ///
+    /// If the underlying value does not implement [`Reflect`]
+    /// or is not of type `T`, returns [`None`].
+    pub fn try_downcast_ref<T: Reflect>(&self) -> Option<&T> {
+        self.as_full()?.downcast_ref()
+    }
+
+    /// Downcasts the value to type `T` by mutable reference.
+    ///
+    /// If the underlying value does not implement [`Reflect`]
+    /// or is not of type `T`, returns [`None`].
+    pub fn try_downcast_mut<T: Reflect>(&mut self) -> Option<&mut T> {
+        self.as_full_mut()?.downcast_mut()
+    }
+
+    /// Returns `true` if the underlying value represents a value of type `T`, or `false`
+    /// otherwise.
+    ///
+    /// Read `is` for more information on underlying values and represented types.
+    #[inline]
+    pub fn represents<T: Reflect>(&self) -> bool {
+        self.type_name() == any::type_name::<T>()
+    }
+}
+
```

Comment:
`represents` makes sense to be here.

originally, yes, the `try_*` methods were stop gaps, to be replaced in unique reflect phase 2.

there are three main places where we use downcasting from `PartialReflect` and i'm pretty sure we can work around the usage everywhere:
1. the registry: i'm confident the registry should be using `Reflect` over `PartialReflect` but i've left the change out of this PR since i reckon the diff would be quite large.
2. tests: many tests downcast trait objects to quickly verify their values, we could add an `assert_reflect_eq` macro that uses `PartialReflect::partial_eq`/secretly downcasts or just bite the readability costs with `.as_full().unwrap().downcast_ref::<T>()`.
3. assorted methods with a `PartialReflect` trait object parameter.
  - `reflect_partial_eq` is currently very awkward. there are distributed efforts to make it friendlier (#6601, #7317) and unique reflect phase 2 will probably involve moving it onto the `Reflect` trait which alleviates the pain around calling `downcast_ref` when the `#[reflect(PartialEq)]` attribute is used.
  - `set` accepts `Box<dyn PartialReflect>` rather than `Box<dyn Reflect>` so it uses `try_take`. frankly, i'm not sure what `set` is used for; all the implementations are almost exactly the same. should the parameter be changed to `Box<dyn Reflect>`?

Reply:
lol, amazingly, the RFC leaves out the type signature of `set`.

Yeah, it should accept `Box<dyn Reflect>` since the type would never match on a proxy type, so it's in practice useless for the types that are `PartialReflect` but not `Reflect`.

I've not considered this before, but you can think of the Unique Reflect RFC as a way to move the "is the underlying type the canonical type" check into the type system. So that when you accept a `dyn Reflect` you say "you must check that the underlying type is canonical". And when you return a `dyn Reflect` you are saying to the consumer that you've already checked that it is a canonical type.

Also: I agree `represents` makes sense here.